libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxSpeechLoader.kt

```diff
@@ -17,28 +20,88 @@ import com.mapbox.navigation.ui.utils.internal.resource.ResourceLoader
     private val options: MapboxSpeechApiOptions,
-    private val resourceLoader: ResourceLoader
-) {
+    private val resourceLoader: ResourceLoader,
+) : VoiceInstructionsDownloadTriggerObserver {
 
-    suspend fun load(typeAndAnnouncement: TypeAndAnnouncement): Expected<Throwable, ByteArray> {
+    private val currentRequests = mutableSetOf<Long>()
+    private val downloadedInstructions = mutableSetOf<TypeAndAnnouncement>()
+    private val downloadedInstructionsLock = Any()
+    private val defaultScope = InternalJobControlFactory.createDefaultScopeJobControl().scope
+
+    suspend fun load(voiceInstruction: VoiceInstructions): Expected<Throwable, ByteArray> {
         return runCatching {
-            val url = instructionUrl(typeAndAnnouncement.announcement, typeAndAnnouncement.type)
-            val response = resourceLoader.load(url)
+            val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction)
+                .getValueOrElse { throw it }
+            val request = createRequest(typeAndAnnouncement).apply {
+                networkRestriction = NetworkRestriction.DISALLOW_ALL
+            }
+            val response = resourceLoader.load(request)
             return processResponse(response)
         }.getOrElse {
             createError(it)
         }
     }
 
-    private suspend fun ResourceLoader.load(url: String) = load(ResourceLoadRequest(url))
+    override fun trigger(voiceInstructions: List<VoiceInstructions>) {
+        defaultScope.launch {
+            voiceInstructions.forEach { voiceInstruction ->
+                val typeAndAnnouncement = VoiceInstructionsParser.parse(voiceInstruction).value
+                if (typeAndAnnouncement != null && !hasTypeAndAnnouncement(typeAndAnnouncement)) {
+                    predownload(typeAndAnnouncement)
+                }
+            }
+        }
+    }
+
+    fun cancel() {
+        defaultScope.cancel()
+        currentRequests.forEach { resourceLoader.cancel(it) }
+    }
+
+    private fun hasTypeAndAnnouncement(typeAndAnnouncement: TypeAndAnnouncement): Boolean {
+        synchronized(downloadedInstructionsLock) {
+            return typeAndAnnouncement in downloadedInstructions
+        }
+    }
+
+    private fun predownload(typeAndAnnouncement: TypeAndAnnouncement) {
+        try {
+            val request = createRequest(typeAndAnnouncement)
+            var id: Long? = null
+            id = resourceLoader.load(request) { result ->
+                id?.let { currentRequests.remove(it) }
+                // tilestore thread
+                if (result.isValue) {
+                    synchronized(downloadedInstructionsLock) {
+                        downloadedInstructions.add(typeAndAnnouncement)
+                    }
+                }
+            }
+            currentRequests.add(id)
+        } catch (ex: Throwable) {
+            logE("Failed to download instruction '$typeAndAnnouncement': ${ex.localizedMessage}")
+        }
+    }
```

Comment:
You wouldn't need to keep `currentRequests` if `predownload` was suspend function like this:
```suggestion
private suspend fun predownload(typeAndAnnouncement: TypeAndAnnouncement) {
        try {
            suspendCancellableCoroutine<Unit> { continuation ->
                val request = createRequest(typeAndAnnouncement)
                var id: Long? = null
                id = resourceLoader.load(request) { result ->
                    // tilestore thread
                    if (result.isValue) {
                        continuation.resume(Unit)
                    }
                }
                continuation.invokeOnCancellation {
                    resourceLoader.cancel(id)
                }
            }
            synchronized(downloadedInstructionsLock) {
                downloadedInstructions.add(typeAndAnnouncement)
            }
        } catch (ex: Throwable) {
            logE("Failed to download instruction '$typeAndAnnouncement': ${ex.localizedMessage}")
        }
    }
```

Reply:
But then the requests won't be run in parallel?